![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# Proyecto: Innovación Tecnológica en Bodoque Bank

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.






# Proyecto

### Equipo:

- Gabriela Mora
- Raúl Silva


### Link de repositorio de GitHub: https://github.com/RaulSilvaA/MDS7202-mora-silva




## 1. Introducción

La predicción de fraude en las empresas en un problema conocido y bastante estudiado debido al interés que supone sobre minizar los riesgos que adoptan las empresas con sus clientes. El problema que se presenta en el proyecto consta de detección de fraude de cliente por parte de banco. 

La forma de abordar el problema consistirá de emplear modelos de clasificación (donde 1 correspondería a cliente clasificado como fraudulento y 0 como cliente no fraudulento) en los diferentes segmentos etarios de la empresa. Los cuales son 4:

* Joven (Menor a 18 años)
* Adulto-Joven (18 - 26 años)
* Adulto (27- 59 años)
* Persona Mayor (60 años o mas).

Donde los datos de entrada del modelo corresponden a datos asociados al cliente donde algunas columnas tienen datos nulos y sus representaciones corresponden a valores negativos, -1 o valores NA. 

En cuanto a la hipótesis del problema, se usará un enfoque donde el aprendizaje de máquinas tiene todo el poder predictivo, o sea en otras palabras, se le entregarán todas las variables que cuenta el dataset y el modelo de ML será el que le de el peso a las variables que son más importantes para predecir este problema. En cuanto a los resultados esperados, no se espera que sean muy ambiciosos debido a que es un típico problema donde hay mucho desbalance de clases y los modelos tenderán a predecir la clase mayoritaria.

## 2. Lectura y Manejo de Datos




### 2.1 Cargar datos en dataframes

In [1]:
#Conectamos el colab a drive para cargar los archivos
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
#Importamos las bases de datos
import pandas as pd

df1=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_1.parquet', engine='pyarrow')
df2=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_2.parquet', engine='pyarrow')
df_email_phone=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_email_phone.parquet', engine='pyarrow')

### 2.2 Exploración inicial

In [3]:
print('Las dimensiones de df_1 son: '+str(df1.shape[0])+' filas por '+str(df1.shape[1])+' columnas')
print('')
print('Las dimensiones de df_2 son: '+str(df2.shape[0])+' filas por '+str(df2.shape[1])+' columnas')
print('')
print('Las dimensiones de df_email_phone son: '+str(df_email_phone.shape[0])+' filas por '+str(df_email_phone.shape[1])+' columnas')

Las dimensiones de df_1 son: 389782 filas por 29 columnas

Las dimensiones de df_2 son: 747410 filas por 29 columnas

Las dimensiones de df_email_phone son: 934730 filas por 7 columnas


Dado el código anterior, se sabe que el primer dataframe df_1 tiene 389782 filas y 29 columnas, el dataframe df_2 tiene 747410 filas y 29 columnas y por último, el dataframe df_email_phone tiene 934730 filas y 7 columnas.

In [4]:
df1.head(5)

id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]

In [5]:
df2.head(5)

id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]

En cuanto a las características que tienen los dataframe similares df_1 y df_2 se puede ver que tienen diferentes variables asociadas al cliente y también si es que el cliente hizo fraude o no. Estas se enumeran a continuación en conjunto a su significado:

* income (numeric):  Ingreso anual del solicitante (en forma decil). Rangos entre [0.1, 0.9].


* prev_address_months_count (numeric): Número de meses en el domicilio registrado anterior del solicitante, es decir, la residencia anterior del solicitante, si corresponde. Varía entre [−1, 380] meses (-1 es un valor faltante).

* current_address_months_count (numeric): meses en la dirección registrada actual del solicitante. Varía entre [−1, 429] meses (-1 es un valor faltante).

* customer_age (numeric): Edad del solicitante en años, redondeada a la década. Rangos entre [10, 90] años.

* days_since_request (numeric): N número de días transcurridos desde que se realizó la solicitud. Rangos entre [0, 79] días.

* intended_balcon_amount (numeric):  Monto inicial transferido para la aplicación. Varía entre [−16, 114] (los negativos son valores faltantes).

* payment_type (categorical): tipo de plan de pago de crédito. 5 valores posibles (anonimizados).

* zip_count_4w (numeric): Number of applications within same zip code in last 4 weeks.
Ranges between [1, 6830].

* velocity_6h (numeric): Velocidad de aplicaciones totales realizadas en las últimas 6 horas, es decir, número promedio de aplicaciones por hora en las últimas 6 horas. Varía entre [−175, 16818].

* velocity_24h (numeric): Velocidad de aplicaciones totales realizadas en las últimas 24 horas, es decir, número promedio de aplicaciones por hora en las últimas 24 horas. Rangos entre [1297, 9586]

* velocity_4w (numeric):  Velocidad de aplicaciones totales realizadas en las últimas 4 semanas, es decir, promedio de aplicaciones por hora en las últimas 4 semanas. Varía entre [2825, 7020].

* bank_branch_count_8w (numeric):  Número de solicitudes totales en la sucursal bancaria seleccionada en las últimas 8 semanas. Varía entre [0, 2404].


* employment_status (categorical):  Situación laboral del solicitante. 7 valores posibles (anonimizados).

* credit_risk_score (numeric): Puntuación interna del riesgo de la aplicación. Varía entre [−191, 389].


* housing_status (categorical): Estado residencial actual del solicitante. 7 valores posibles (anonimizados).


* bank_months_count (numeric): ¿Cuántos años tiene la cuenta anterior (si se tiene) en meses? Varía entre [−1, 32] meses (-1 es un valor faltante).

* has_other_cards (binary): Si el solicitante tiene otras tarjetas de la misma compañía bancaria.

* proposed_credit_limit (numeric): límite de crédito propuesto por el solicitante. Rangos entre [200, 2000].

* foreign_request (binary): si el país de origen de la solicitud es diferente del país del banco.

* source (categorical): Fuente en línea de aplicación. Ya sea navegador (INTERNET) o app (TELEAPP).

* session_length_in_minutes (numeric):  duración de la sesión del usuario en el sitio web bancario en minutos. Varía entre [−1, 107] minutos (-1 es un valor faltante).

* device_os (categorical): Sistema operativo del dispositivo que realizó la solicitud. Los valores posibles son: Windows, macOS, Linux, X11 u otros.

* keep_alive_session (binary): opción de usuario al cerrar la sesión.


* device_fraud_count (numeric):  Número de aplicaciones fraudulentas con dispositivo usado. Rangos entre [0, 1].

* month (numeric): Mes en que se realizó la solicitud. Rangos entre [0, 7].

* fraud_bool (binary): Si la aplicación es fraudulenta o no.

In [6]:
df_email_phone.head(5)

id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1

Y en cuanto a la base de datos df_email_phone, las columnas que tiene son:
* name_email_similarity (numeric): Métrica de similitud entre el correo electrónico y el nombre del solicitante. Los valores más altos representan una mayor similitud. Rangos entre [0, 1].
* date_of_birth_distinct_emails_4w (numeric): Número de correos electrónicos para solicitantes con la misma fecha de nacimiento en las últimas 4 semanas. Rangos entre [0, 39].
* email_is_free (binary): Dominio del correo electrónico de la aplicación (ya sea gratuito o de pago).
* phone_home_valid (binary): Validez del teléfono de casa proporcionado.
* phone_mobile_valid (binary): Validez del teléfono móvil proporcionado.
* device_distinct_emails (numeric): Número de correos electrónicos distintos en el sitio web bancario desde el dispositivo usado en las últimas 8 semanas. Varía entre [−1, 2] correos electrónicos (-1 es un valor faltante).

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 389782 entries, 104446 to 193209
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            389782 non-null  int64  
 1   fraud_bool                    389782 non-null  int64  
 2   income                        389782 non-null  float64
 3   prev_address_months_count     389782 non-null  int64  
 4   current_address_months_count  389782 non-null  int64  
 5   customer_age                  389782 non-null  int64  
 6   days_since_request            389782 non-null  float64
 7   intended_balcon_amount        389782 non-null  float64
 8   payment_type                  389782 non-null  object 
 9   zip_count_4w                  389782 non-null  int64  
 10  velocity_6h                   389782 non-null  float64
 11  velocity_24h                  389782 non-null  float64
 12  velocity_4w                   389782 no

In [8]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 747410 entries, 699954 to 384611
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            747410 non-null  int64  
 1   fraud_bool                    747410 non-null  int64  
 2   income                        747410 non-null  float64
 3   prev_address_months_count     747410 non-null  int64  
 4   current_address_months_count  747410 non-null  int64  
 5   customer_age                  747410 non-null  int64  
 6   days_since_request            747410 non-null  float64
 7   intended_balcon_amount        747410 non-null  float64
 8   payment_type                  747410 non-null  object 
 9   zip_count_4w                  747410 non-null  int64  
 10  velocity_6h                   747410 non-null  float64
 11  velocity_24h                  747410 non-null  float64
 12  velocity_4w                   747410 no

In [9]:
df_email_phone.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 934730 entries, 987231 to 986114
Data columns (total 7 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                934730 non-null  int64  
 1   name_email_similarity             934730 non-null  float64
 2   date_of_birth_distinct_emails_4w  934730 non-null  int64  
 3   email_is_free                     934730 non-null  int64  
 4   device_distinct_emails_8w         934730 non-null  int64  
 5   phone_home_valid                  934730 non-null  int64  
 6   phone_mobile_valid                934730 non-null  int64  
dtypes: float64(1), int64(6)
memory usage: 57.1 MB


### 2.3 Unir dataframes

In [10]:
#Se agrega una columna para tener la trazabilidad de los orígenes de los datos
df1['origen']='base 1'
df2['origen']='base 2'

In [11]:
df_concat=pd.concat([df1,df2]).reset_index(drop=True)

In [12]:
len(df_concat.drop_duplicates())

973900

### 2.4 Unir nuevas variables

In [13]:
bbdd_final = pd.merge(df_concat, df_email_phone, on='id', how='left')

In [14]:
bbdd_final.shape

(1137192, 36)

In [15]:
bbdd_final.groupby('device_os').size()

device_os
linux        383185
macintosh     57016
other        341352
windows      346379
x11            9260
dtype: int64

Al unir las variables se verifica cómo se ven los valores de las columnas con el objetivo de ver algún valor extraño. Primero se verán las columans numericas y luego las categoricas,

In [16]:
bbdd_final.describe().transpose()

count           mean            std  \
id                                1137192.0  500113.268421  288722.419048   
fraud_bool                        1137192.0       0.011010       0.104347   
income                            1137192.0       0.578548       0.288240   
prev_address_months_count         1137192.0      14.647000      42.981190   
current_address_months_count      1137192.0      99.175138      94.063764   
customer_age                      1137192.0      41.343828      13.766449   
days_since_request                1137192.0       0.901719       5.000559   
intended_balcon_amount            1137192.0       8.560989      20.532602   
zip_count_4w                      1137192.0    1518.532112     966.147032   
velocity_6h                       1137192.0    5491.808343    2940.041218   
velocity_24h                      1137192.0    4661.668206    1451.255984   
velocity_4w                       1137192.0    4734.725941     871.270566   
bank_branch_count_8w              1137192.0     201.271150     473.813362   
credit_risk_score                 1137192.0     139.263714      71.452421   
bank_months_count                 1137192.0      11.148524      12.128297   
has_other_cards                   1137192.0       0.249646       0.432809   
proposed_credit_limit             1137192.0     551.614890     506.588620   
foreign_request                   1137192.0       0.023880       0.152675   
session_length_in_minutes         1137192.0       7.817620       8.242504   
keep_alive_session                1137192.0       0.556244       0.496827   
device_fraud_count                1137192.0       0.000000       0.000000   
month                             1137192.0       3.656215       2.116896   
x1                                1137192.0       0.014586       1.013109   
x2                                1137192.0       0.011012       1.012876   
name_email_similarity             1057465.0       0.487641       0.291457   
date_of_birth_distinct_emails_4w  1057465.0       7.778368       4.814321   
email_is_free                     1057465.0       0.518114       0.499672   
device_distinct_emails_8w         1057465.0       1.022282       0.191573   
phone_home_valid                  1057465.0       0.493414       0.499957   
phone_mobile_valid                1057465.0       0.856897       0.350178   

                                           min            25%            50%  \
id                                0.000000e+00  249918.500000  500347.500000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                            1.000000e-01       0.300000       0.600000   
prev_address_months_count        -1.000000e+00      -1.000000      -1.000000   
current_address_months_count     -1.000000e+00      27.000000      64.000000   
customer_age                      1.000000e+01      30.000000      50.000000   
days_since_request                3.112791e-08       0.007434       0.015659   
intended_balcon_amount           -1.553733e+01      -1.179949      -0.834103   
zip_count_4w                      1.000000e+00     886.000000    1209.000000   
velocity_6h                      -1.741097e+02    3337.138253    5192.389354   
velocity_24h                      1.322325e+03    3504.029907    4641.080855   
velocity_4w                       2.870592e+03    4238.872451    4813.757967   
bank_branch_count_8w              0.000000e+00       1.000000      10.000000   
credit_risk_score                -1.770000e+02      90.000000     130.000000   
bank_months_count                -1.000000e+00       1.000000       6.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit             1.900000e+02     200.000000     200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes        -1.000000e+00       3.152311       5.248896   
keep_alive_session                0.000000e+00    

de las variables numericas se puede observar que la variable device_fraud_count no tiene variabilidad por lo que se debe eliminar

In [17]:
bbdd_final.describe(include = [object]).transpose()

count unique       top     freq
payment_type       1137192      5        AB   453427
employment_status  1137192      7        CA   778603
housing_status     1137192      7        BC   381659
source             1137192      2  INTERNET  1128411
device_os          1137192      5     linux   383185
origen             1137192      2    base 2   747410

In [18]:
print(bbdd_final['payment_type'].unique())
print(bbdd_final['employment_status'].unique())
print(bbdd_final['source'].unique())
print(bbdd_final['device_os'].unique())
print(bbdd_final['housing_status'].unique())

['AC' 'AA' 'AB' 'AD' 'AE']
['CC' 'CF' 'CA' 'CB' 'CE' 'CD' 'CG']
['INTERNET' 'TELEAPP']
['windows' 'linux' 'other' 'macintosh' 'x11']
['BB' 'BA' 'BE' 'BC' 'BD' 'BF' 'BG']


Y en cuanto a las variables categoricas se observa que hay valores normales en su data.

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



#### Nulos

Primero se transformarán los valores nulos respectivos por cada columna donde la forma de representar los nulos correspondía a -1 o a valores negativos, luego se cuenta la cantidad de nulos que hay por columna. Se puede observar que las columnas que se agregaron por el cruce son las que tienen estos valores nulos, además de las columnas, prev_address_months_count, intended_balcon_amount, current_address_months_count, bank_months_count, session_length_in_minutes, payment_tyoe, employment_status, source, device_os y housing_status.

In [19]:
import numpy as np
#transformar a valores nulos

bbdd_final['prev_address_months_count']=bbdd_final['prev_address_months_count'].replace(-1, np.NaN) 
bbdd_final['current_address_months_count']=bbdd_final['current_address_months_count'].replace(-1, np.NaN) 
bbdd_final.intended_balcon_amount = np.where(bbdd_final.intended_balcon_amount < 0, 0, bbdd_final.intended_balcon_amount)
bbdd_final['intended_balcon_amount']=bbdd_final['intended_balcon_amount'].replace(0, np.NaN)
bbdd_final['bank_months_count']=bbdd_final['bank_months_count'].replace(-1, np.NaN) 
bbdd_final['session_length_in_minutes']=bbdd_final['session_length_in_minutes'].replace(-1, np.NaN) 
bbdd_final['device_distinct_emails_8w']=bbdd_final['device_distinct_emails_8w'].replace(-1, np.NaN)   


In [20]:
bbdd_final.shape

(1137192, 36)

In [21]:
bbdd_final.isnull().sum()

id                                       0
fraud_bool                               0
income                                   0
prev_address_months_count           866476
current_address_months_count          3925
customer_age                             0
days_since_request                       0
intended_balcon_amount              854466
payment_type                             0
zip_count_4w                             0
velocity_6h                              0
velocity_24h                             0
velocity_4w                              0
bank_branch_count_8w                     0
employment_status                        0
credit_risk_score                        0
housing_status                           0
bank_months_count                   280821
has_other_cards                          0
proposed_credit_limit                    0
foreign_request                          0
source                                   0
session_length_in_minutes             2574
device_os  

Lo que se realizará por cada columna con nulos es lo siguiente:


1.   Para variables que tienen muchas columnas con nulos, estas se eliminan: caso de prev_address_months_count y de intended_balcon_amount, payment_tyoe, employment_status, source, device_os y housing_status.
2.   Para las últimas columnas agregadas por el cruce, se investigará el origen de los nulos.
3. Para el resto de las  columnas, se eligirá la media como valor para imputar esos valores.








**De aquí en adelante se investigará el origen de las últimas columnas:**
Por la cantidad repetida de nulos en cada columna, una hipótesis es que todos esos valores nulos provienen de los mismos registros. Esto se comprueba a continuación, donde primero se toma una columna cualquiera, luego se filtra por las filas nulas y se vuelve a calcular la cantidad de nulos que hay por cada columna. Si da la misma cantidad, esto significa que los nulos provienen de los mismos registros.

In [22]:
nulos=bbdd_final[bbdd_final['name_email_similarity'].isnull()]

In [23]:
nulos.isnull().sum()

id                                      0
fraud_bool                              0
income                                  0
prev_address_months_count           60617
current_address_months_count          290
customer_age                            0
days_since_request                      0
intended_balcon_amount              60042
payment_type                            0
zip_count_4w                            0
velocity_6h                             0
velocity_24h                            0
velocity_4w                             0
bank_branch_count_8w                    0
employment_status                       0
credit_risk_score                       0
housing_status                          0
bank_months_count                   19719
has_other_cards                         0
proposed_credit_limit                   0
foreign_request                         0
source                                  0
session_length_in_minutes             199
device_os                         

Como dio la misma cantidad de nulos que en la base donde no se filtro, quiere decir que todos los nulos provienen de los mismos registros.

Para saber la razón de los nulos, se separá la base en una que contiene sólo los registros con nulos y la otra en una base que sólo contenga datos sin nulos, luego de esta separación se verificarán 2 cosas. Primero, ver si hay diferencias notables en los estadísticos de ambas bases. Esto con el objetivo de ver si hay alguna diferencia en las poblaciones de ambas bases que pueda llevar a pensar que hay una variable que se correlaciona con la nulidad de algunos datos.

Lo segundo que se verificará es si el origen de los datos puede llegar a afectar la nulidad, debido a que df1 y df2 tienen 2 orígenes distintos y además, los datos de df2 no son tan confiables, según enunciado, pero no sabemos el por qué.

##### Estadísticos

Ya habíamos separado una base solo con nulos, así que falta armar la base sin nulos. Luego de crearla, se verán las diferencias entre los estadísticos de ambas bases.

In [24]:
sin_nulos=bbdd_final[bbdd_final['name_email_similarity'].notnull()]

In [25]:
sin_nulos.describe().transpose()

count           mean            std  \
id                                1057465.0  500141.823877  288698.160011   
fraud_bool                        1057465.0       0.011076       0.104660   
income                            1057465.0       0.578704       0.288263   
prev_address_months_count          251606.0      64.766595      66.861254   
current_address_months_count      1053830.0      99.517182      94.010354   
customer_age                      1057465.0      41.342626      13.764593   
days_since_request                1057465.0       0.901141       4.994621   
intended_balcon_amount             263041.0      37.487422      24.101325   
zip_count_4w                      1057465.0    1518.330710     966.003968   
velocity_6h                       1057465.0    5492.639968    2939.143965   
velocity_24h                      1057465.0    4661.051132    1450.642353   
velocity_4w                       1057465.0    4734.927085     871.015792   
bank_branch_count_8w              1057465.0     200.955029     473.432868   
credit_risk_score                 1057465.0     139.279955      71.445968   
bank_months_count                  796363.0      15.130646      11.447759   
has_other_cards                   1057465.0       0.249702       0.432841   
proposed_credit_limit             1057465.0     551.541668     506.698808   
foreign_request                   1057465.0       0.023902       0.152742   
session_length_in_minutes         1055090.0       7.836127       8.240376   
keep_alive_session                1057465.0       0.556186       0.496833   
device_fraud_count                1057465.0       0.000000       0.000000   
month                             1057465.0       3.656299       2.117072   
x1                                1057465.0       0.014140       1.013339   
x2                                1057465.0       0.011147       1.012887   
name_email_similarity             1057465.0       0.487641       0.291457   
date_of_birth_distinct_emails_4w  1057465.0       7.778368       4.814321   
email_is_free                     1057465.0       0.518114       0.499672   
device_distinct_emails_8w         1057145.0       1.022894       0.188342   
phone_home_valid                  1057465.0       0.493414       0.499957   
phone_mobile_valid                1057465.0       0.856897       0.350178   

                                           min            25%            50%  \
id                                0.000000e+00  249877.000000  500452.000000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                            1.000000e-01       0.300000       0.600000   
prev_address_months_count         6.000000e+00      26.000000      36.000000   
current_address_months_count      0.000000e+00      27.000000      65.000000   
customer_age                      1.000000e+01      30.000000      50.000000   
days_since_request                3.112791e-08       0.007444       0.015666   
intended_balcon_amount            2.738965e-05      20.767272      32.973873   
zip_count_4w                      1.000000e+00     886.000000    1208.000000   
velocity_6h                      -1.741097e+02    3339.402765    5193.504274   
velocity_24h                      1.322325e+03    3504.200445    4639.960378   
velocity_4w                       2.870592e+03    4238.832450    4813.957271   
bank_branch_count_8w              0.000000e+00       1.000000      10.000000   
credit_risk_score                -1.770000e+02      90.000000     130.000000   
bank_months_count                 1.000000e+00       2.000000      15.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit             1.900000e+02     200.000000     200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes         4.088612e-05       3.166989       5.257871   
keep_alive_session                0.000000e+00    

In [26]:
nulos.describe().transpose()

count           mean            std  \
id                                79727.0  499734.521003  289045.533576   
fraud_bool                        79727.0       0.010122       0.100099   
income                            79727.0       0.576482       0.287923   
prev_address_months_count         19110.0      64.220984      66.661223   
current_address_months_count      79437.0      99.587170      94.450984   
customer_age                      79727.0      41.359765      13.791121   
days_since_request                79727.0       0.909396       5.078683   
intended_balcon_amount            19685.0      37.649522      24.042736   
zip_count_4w                      79727.0    1521.203432     968.044676   
velocity_6h                       79727.0    5480.778022    2951.912548   
velocity_24h                      79727.0    4669.852812    1459.354979   
velocity_4w                       79727.0    4732.058053     874.643868   
bank_branch_count_8w              79727.0     205.464046     478.814725   
credit_risk_score                 79727.0     139.048302      71.538052   
bank_months_count                 60008.0      15.153746      11.456034   
has_other_cards                   79727.0       0.248912       0.432385   
proposed_credit_limit             79727.0     552.586075     505.127016   
foreign_request                   79727.0       0.023593       0.151778   
session_length_in_minutes         79528.0       7.857477       8.251138   
keep_alive_session                79727.0       0.557013       0.496742   
device_fraud_count                79727.0       0.000000       0.000000   
month                             79727.0       3.655098       2.114583   
x1                                79727.0       0.020506       1.010037   
x2                                79727.0       0.009233       1.012735   
name_email_similarity                 0.0            NaN            NaN   
date_of_birth_distinct_emails_4w      0.0            NaN            NaN   
email_is_free                         0.0            NaN            NaN   
device_distinct_emails_8w             0.0            NaN            NaN   
phone_home_valid                      0.0            NaN            NaN   
phone_mobile_valid                    0.0            NaN            NaN   

                                           min            25%            50%  \
id                                3.700000e+01  250654.500000  498688.000000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                            1.000000e-01       0.300000       0.600000   
prev_address_months_count         6.000000e+00      26.000000      35.000000   
current_address_months_count      0.000000e+00      27.000000      64.000000   
customer_age                      1.000000e+01      30.000000      50.000000   
days_since_request                3.839882e-07       0.007316       0.015572   
intended_balcon_amount            7.248630e-04      21.080925      32.996352   
zip_count_4w                      4.000000e+00     886.000000    1212.000000   
velocity_6h                       1.727520e+01    3304.988517    5175.170149   
velocity_24h                      1.328306e+03    3501.726907    4657.212140   
velocity_4w                       2.961316e+03    4239.241432    4810.641928   
bank_branch_count_8w              0.000000e+00       1.000000      10.000000   
credit_risk_score                -1.480000e+02      89.000000     129.000000   
bank_months_count                 1.000000e+00       2.000000      15.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit             1.900000e+02     200.000000     200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes         1.178154e-02       3.196885       5.265800   
keep_alive_session                0.000000e+00       0.000000       1.000000   
device_fraud_count            

In [27]:
resta1=sin_nulos.describe().transpose()

In [28]:
resta2=nulos.describe().transpose()

In [29]:
resta1-resta2

count        mean         std  \
id                                 977738.0  407.302874 -347.373566   
fraud_bool                         977738.0    0.000954    0.004562   
income                             977738.0    0.002221    0.000340   
prev_address_months_count          232496.0    0.545612    0.200031   
current_address_months_count       974393.0   -0.069988   -0.440630   
customer_age                       977738.0   -0.017139   -0.026528   
days_since_request                 977738.0   -0.008255   -0.084062   
intended_balcon_amount             243356.0   -0.162100    0.058589   
zip_count_4w                       977738.0   -2.872722   -2.040708   
velocity_6h                        977738.0   11.861946  -12.768583   
velocity_24h                       977738.0   -8.801680   -8.712626   
velocity_4w                        977738.0    2.869032   -3.628076   
bank_branch_count_8w               977738.0   -4.509017   -5.381857   
credit_risk_score                  977738.0    0.231653   -0.092084   
bank_months_count                  736355.0   -0.023100   -0.008275   
has_other_cards                    977738.0    0.000790    0.000455   
proposed_credit_limit              977738.0   -1.044407    1.571792   
foreign_request                    977738.0    0.000308    0.000964   
session_length_in_minutes          975562.0   -0.021350   -0.010762   
keep_alive_session                 977738.0   -0.000828    0.000091   
device_fraud_count                 977738.0    0.000000    0.000000   
month                              977738.0    0.001201    0.002488   
x1                                 977738.0   -0.006366    0.003302   
x2                                 977738.0    0.001913    0.000152   
name_email_similarity             1057465.0         NaN         NaN   
date_of_birth_distinct_emails_4w  1057465.0         NaN         NaN   
email_is_free                     1057465.0         NaN         NaN   
device_distinct_emails_8w         1057145.0         NaN         NaN   
phone_home_valid                  1057465.0         NaN         NaN   
phone_mobile_valid                1057465.0         NaN         NaN   

                                           min         25%          50%  \
id                               -3.700000e+01 -777.500000  1764.000000   
fraud_bool                        0.000000e+00    0.000000     0.000000   
income                            0.000000e+00    0.000000     0.000000   
prev_address_months_count         0.000000e+00    0.000000     1.000000   
current_address_months_count      0.000000e+00    0.000000     1.000000   
customer_age                      0.000000e+00    0.000000     0.000000   
days_since_request               -3.528603e-07    0.000128     0.000094   
intended_balcon_amount           -6.974734e-04   -0.313653    -0.022479   
zip_count_4w                     -3.000000e+00    0.000000    -4.000000   
velocity_6h                      -1.913849e+02   34.414248    18.334124   
velocity_24h                     -5.980949e+00    2.473538   -17.251762   
velocity_4w                      -9.072466e+01   -0.408982     3.315343   
bank_branch_count_8w              0.000000e+00    0.000000     0.000000   
credit_risk_score                -2.900000e+01    1.000000     1.000000   
bank_months_count                 0.000000e+00    0.000000     0.000000   
has_other_cards                   0.000000e+00    0.000000     0.000000   
proposed_credit_limit             0.000000e+00    0.000000     0.000000   
foreign_request                   0.000000e+00    0.000000     0.000000   
session_length_in_minutes        -1.174065e-02   -0.029896    -0.007930   
keep_alive_session                0.000000e+00    0.000000     0.000000   
device_fraud_count                0.000000e+00    0.000000     0.000000   
month                             0.000000e+00    0.000000     0.000000   
x1                               -5.774191e-01   -0.013670    -0.009219   
x2                               -3.

Al ver las diferencias que pueden haber entre los estadísticos de ambas bases, se puede ver que las mayores diferencias se dan en las columnas: zip_count_4w, velocity_6h, velocity_24h, velocity_4w y bank_branch_count_8w. Al observar los valores de esas columnas en los estadísticos de las bases respectivas, se puede observar que la magnitud más alta en comparación a las otras columnas se debe al rango en donde se mueven esas variables, siendo estos mucho más amplios que el resto de las variables. 
Por lo que se concluye que no hay diferencias significativas entre ambos estadisticos. Por lo que se explorará si el origen de los nulos proviene de los orígenes de las bases.

##### Origen

In [30]:
nulos.groupby(['origen'])['origen'].count()

origen
base 2    79727
Name: origen, dtype: int64

In [31]:
sin_nulos.groupby(['origen'])['origen'].count()

origen
base 1    389782
base 2    667683
Name: origen, dtype: int64

Se puede observar que los datos nulos provienen solo del origen de la base 2. Esto se puede deber a la confiabilidad de los datos que se informa en el inicio del proyecto, donde mencionan la confiabilidad de ambas bases: **"la primera más confiable que la segunda"**. No mencionan más sobre la confiabilidad de los datos, pero con la información que tenemos se puede suponer que la confiabilidad de los datos de la segunda base puede tener relación a que en ella no se pudieron recolectar datos de contacto del cliente por un motivo aleatorio, esto debido a que no se observó que alguna variable extra del dataset tuviera relación con la nulidad de los datos.

Como este tipo de nulo es un missing at random, la solución más óptima es dropear los registros que contengan los nulos.

#### Duplicados


Primero se verá cuántos datos son los duplicados en la base y luego, mostrarán los casos que se repiten en la base de datos

In [32]:
bbdd_final=bbdd_final.drop(['origen'], axis=1)
len(bbdd_final[bbdd_final.duplicated(keep=False)])

299655

In [33]:
pd.set_option('display.max_columns', None)

In [34]:
bbdd_final[bbdd_final.duplicated(keep=False)]

id  fraud_bool  income  prev_address_months_count  \
0        729517           0     0.7                        NaN   
1        149585           0     0.8                        NaN   
3        825283           0     0.5                        NaN   
4          8308           0     0.9                        NaN   
5        891636           0     0.6                        NaN   
...         ...         ...     ...                        ...   
1137177  265707           0     0.2                        NaN   
1137180  690720           0     0.6                       31.0   
1137187   28360           0     0.8                        NaN   
1137188  428637           0     0.6                        NaN   
1137190  656236           0     0.6                        NaN   

         current_address_months_count  customer_age  days_since_request  \
0                               305.0            60            0.030059   
1                               140.0            50            0.015659   
3                                85.0            30            0.027292   
4                                39.0            30            0.010945   
5                               137.0            60            0.046158   
...                               ...           ...                 ...   
1137177                         249.0            70            0.022940   
1137180                          12.0            20            0.016682   
1137187                         137.0            30            0.018466   
1137188                          60.0            50            0.026457   
1137190                          85.0            60            0.008277   

         intended_balcon_amount payment_type  zip_count_4w  velocity_6h  \
0                           NaN           AC           990  5340.131605   
1                      3.951994           AA          1269  2844.260015   
3                           NaN           AB          1698  7938.525709   
4                           NaN           AC           569  5268.975264   
5                     25.841925           AA          1512  7669.494145   
...                         ...          ...           ...          ...   
1137177                     NaN           AD          4174  1925.667232   
1137180               25.274510           AA          1553  1694.710265   
1137187               49.088586           AA           915  4814.292418   
1137188                     NaN           AD          3145  7215.613670   
1137190                     NaN           AB          1021  1247.492130   

         velocity_24h  velocity_4w  bank_branch_count_8w employment_status  \
0         3945.871264  5531.610889                     1                CC   
1         2789.827355  3141.137362                  1605                CF   
3         4963.194619  5070.085151                   872                CA   
4         5573.932997  4931.751648                     1                CA   
5         4251.257682  5721.121146                   271                CC   
...               ...          ...                   ...               ...   
1137177   5056.733128  5723.696856                     8                CB   
1137180   6512.247247  5264.103915                    30                CA   
1137187   3219.381648  5444.343899                    19                CA   
1137188   5311.829165  5318.864311                  1831                CA   
1137190   3224.472911  4409.105459                    18                CE   

         credit_risk_score housing_status  bank_months_count  has_other_cards  \
0                      147             BB                NaN                0   
1                      105             BA                9.0                1   
3                      138             BE                3.0                0   
4                      110             BB                NaN                1   
5                       33             BB               28.0                

Como cada registro representa la situación de un cliente específico, que hayan registros con la misma información es incorrecto y podría inducir a sesgos en el análisis, por lo que la forma de solucionar este problema es dropeando los registros duplicados.

### 2.6 Limpiar

Se eliminan de la base los registros duplicados 
device_fraud_count

In [35]:
bbdd_final=bbdd_final.drop_duplicates(keep='first')

In [36]:
bbdd_final.shape

(934858, 35)

Se eliminan las columnas que tienen muchos NA y la variable device_fraud_count que no presenta variacion

In [37]:
bbdd_final=bbdd_final.drop(['prev_address_months_count','intended_balcon_amount','device_fraud_count'], axis=1)

Se eliminan las filas que tienen las últimas columnas con NA y también el id 0 debido a que parece ser un dato de prueba que no significa nada

In [38]:
bbdd_final=bbdd_final[bbdd_final['name_email_similarity'].notnull()]
bbdd_final = bbdd_final[bbdd_final['id'] != 0]

Además de eliminar esos datos, se eliminan las columnas x1 y x2, ya que no significan nada dentro de la base de datos.

In [39]:
bbdd_final=bbdd_final.drop(['x1','x2'], axis=1)

Por último, las otras varibles que presentaban valores menores de nulos se les imputaran esos valores por la mdedia del resto en caso de ser numerica y en caso de ser categorica, la moda


In [40]:
bbdd_final['current_address_months_count'] = bbdd_final['current_address_months_count'].fillna(bbdd_final['current_address_months_count'].mean())
bbdd_final['bank_months_count'] = bbdd_final['bank_months_count'].fillna(bbdd_final['bank_months_count'].mean())
bbdd_final['session_length_in_minutes'] = bbdd_final['session_length_in_minutes'].fillna(bbdd_final['session_length_in_minutes'].mean())
bbdd_final['device_distinct_emails_8w'] = bbdd_final['device_distinct_emails_8w'].fillna(bbdd_final['device_distinct_emails_8w'].mean().round())                             

In [41]:
bbdd_final.isnull().sum()

id                                  0
fraud_bool                          0
income                              0
current_address_months_count        0
customer_age                        0
days_since_request                  0
payment_type                        0
zip_count_4w                        0
velocity_6h                         0
velocity_24h                        0
velocity_4w                         0
bank_branch_count_8w                0
employment_status                   0
credit_risk_score                   0
housing_status                      0
bank_months_count                   0
has_other_cards                     0
proposed_credit_limit               0
foreign_request                     0
source                              0
session_length_in_minutes           0
device_os                           0
keep_alive_session                  0
month                               0
name_email_similarity               0
date_of_birth_distinct_emails_4w    0
email_is_fre

### 2.7 Nueva columna con segmentaciones

In [42]:
import numpy as np
bins = [0, 18, 26, 59, np.inf]
names = ['Menor a 18', '19 - 26 años', '27 - 59 años', '60 años o mas']

bbdd_final['segmentacion_etaria'] = pd.cut(bbdd_final['customer_age'], bins, labels=names)

In [43]:
bbdd_final.shape

(873770, 31)

In [44]:
bbdd_final[bbdd_final['fraud_bool']>0].describe

<bound method NDFrame.describe of              id  fraud_bool  income  current_address_months_count  \
89       110109           1     0.5                         109.0   
90       204980           1     0.7                          36.0   
146      370826           1     0.8                          84.0   
239      787639           1     0.9                         303.0   
246      154083           1     0.9                          76.0   
...         ...         ...     ...                           ...   
1136870  138943           1     0.8                          66.0   
1136888  308929           1     0.8                         143.0   
1136894  738506           1     0.8                         193.0   
1137104  653096           1     0.1                         106.0   
1137121  686403           1     0.9                         111.0   

         customer_age  days_since_request payment_type  zip_count_4w  \
89                 40            0.033339           AD          1

In [45]:
bbdd_final.shape

(873770, 31)

## 3. Análisis Exploratorio de Datos

In [46]:
!pip install ydata_profiling

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [47]:
from ydata_profiling import ProfileReport

profile = ProfileReport(bbdd_final, title="EDA")
profile.to_file("eda.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

### 3.1 Análisis del EDA

* **id**: ID único de cada usuario.

* **fraud_bool**: Variable categórica que representa si una cuenta es fraudulenta o no. La clase está completamente desbalanceada, pues hay 864148 cuentas que no son fraude, post limpieza, mientras que los frades son solo 9622, representando solo el 1.1% de los datos disponibles.

* **income**: Variable numérica con valores entre 0 y 1. Posee una distribución exponencial, salvo que contiene una leve acumulación con valores entre 0 y 0.1.

* **current_address_months_count**: Variable numérica con distribución que disminuye logarítmicamente. Toma valores entre 0 y 429, con promedio en 99. No posee relaciones aparentes con el resto de las variables

* **customer_age**: Variable numérica positiva con valores entre el 10 y el 90, representando la edad del cliente. Posee una distribución normal de los datos, con media en 41 con desviación standar de 13. Posee relación directa con la segmentación etaria.

* **days_since_request**: Variable numérica con valores positivos y distribución desconocida. Posee un cúmulo muy grande de datos entre 0 y 5, pero un par de dats mayores a este valor, siendo el máximo cercano a 76. Su media se encuentra en 0.9 y no posee relación notable con el resto de las variables. 

* **payment_type**: Variable categórica que toma los valores de AX con X en \[A, E\]. El valor que se ve más representado es la categoría AB con 348704, mientras que AE está presente tan solo 207 clientes. En cuanto a su relación con el resto de las variables, a primera vista solo se aprecia una leve relación con la variable *months_count*

* **zip_count_4w**: Variable numérica con distribución similar a dos normales, con media en 1517 y desviación estándar de 965. Son solo valores enteros positivos que fluctuan entre 0 y 6650. Posee leves relacions con las variables de las velocidades y una relación inversa con la columna *month*.

* **velocity_6h**: Variable numérica con distribución normal con media centrada en 5491, desviación estándar de 	2940 y valores entre 	-174 y 16754. Con respecto a las relaciones, posee las mismas que la variable anterior.

* **velocity_24h**: Variable numérica con distribución desconocida, pero que posee dos puntos donde se acumulan los datos de manera normal. Los valores fluctúan entre 1322 y 9540 a la vez que su relación con el resto de variables es idéntica a la anterior mencionada.

* **velocity_4w**: Variable numérica con distribución desconocida, pero que posee dos puntos donde se acumulan los datos de manera normal. Los valores fluctúan entre 2870 y 7020 a la vez que su relación con el resto de variables es idéntica a la anterior mencionada. Particularmente, posee una relación -1 con la variable *month* mucho más marcada que sus compañeras.

* **bank_branch_count_8w**: Variable numérica con valores positivos entre 0 y 2377. El 305 de los datos son 0 o 1, mientras que el resto se distribuye en mucha menos cantidad para cada valor sobrante. No posee relaciones claras con el resto de columnas.

* **employment_status**: Variable categórica con valores de la forma CX, con X en \[A,G\]. El más común es CA con 598651, rperesentando más de la mitad de la población de análisis, por otro lado CG representa a menos del 1% del mismo. No posee relaciones claras con el resto de columnas.

* **credit_risk_score**: Variable numérica con distribución normal con valores entre -177 y	387 y media en 139.25 con desviación estándar de 71.4. Posee leves interacciones con el resto de variables, pero la más notoria es con *credit_limit*.

* **housing_status**: Se trata de una variable categórica que posee 7 categorías distinta, las que se encuentran codificadas en dos caracteres BX con X entre \[A y G\]. BC es el que posee la mayor frecuencia, representado al 33.5% de los datos, seguido de cerca de BB con un 30.1% de los mismos. Para finalizar, no posee mayores interacciones ni relación con otras variables.

* **bank_months_count**: La variable se contiene entre 1 y 32, sin nulos. El promedio está centrado en 15.17, además de tener la mayor cantidad de datos concentrados entre 12 y 17. El otro cúmulo de datos se encuentra en los valores menores. 

* **has_another_card**: la variable posee 2 posibles valores, 0 y 1, donde cuando toma el valor 0 corresponde a que no tiene otras tarjetas y cuando es 1 significa que sí posee otras tarjetas. Los valores están desbalanceados, donde el valor 1 corresponde a menos de 1/3 de los valores que corresponden a 0.

 En cuanto a las correlaciones con otras variables, no se observa que tenga una correlacion significativa con otras variables, se puede observar que ronda una correlacion cercana a 0. Por último, no hay valores nulos enla variable.

* **proposer_credit_limit**: la variable posee 12 valores distintos, a pesar de estar definida como una variable flotante. En cuanto a su distribución, no se observa una distribución clara de los datos. y casi el 58% de los datos corresponde a la variable 200. Por otro lado, los valores altos de esta variable no parecen ser outliers debido a que presentan cierta frecuencia, esta no es con valores menores a 10. Po último, no se observan valores nulos en estas variables.

 En cuanto a los estadísticos de esta variable, la media ronda por 551, y las medidas de dispersión dan cuenta de una alta dispersión de los datos. Esto tiene sentido mirando el histograma debido a que no se observa que se concentren los datos en algún lugar.

 En cuanto a las correlaciones con otras variables, credit_risk_score presenta el mayor valor de correlacion con esta variable teniendo un orden de 0.6. En cuanto a las otras variables, no pasan de una correlacion de 0.2.

* **foreign_request**: esta variable tiene solo 2 tipos de valores, 0 y 1, donde el valor 1 quiere decir que la solicitud proviene de un pais distinta al banco y cuando es 0, el pais es el mismo del banco. Las clases en esta variable están muy desbalanceadas, donde el valor 0 corresponde a 97,6% de los datos de esta variable. En cuanto a las correlaciones de esta variable, esta no presenta variables correlacionadas distintas a valores cercanos a 0. Tampoco se presentan valores nulos en la variable.

* **source**: esta variable tiene solo 2 valores, internet y teleapp, donde la variable internet corresponde al 99% de los valores de esa variable. En cuanto a las correlaciones de esa variable, la variable device_distinct_emails_8w	tiene una correlación igual a 0.419 freente a las otras variables donde casi son 0. En cuanto a los valores nulos, no se encuentran valores nulos en esta variable.

* **session_length_in_minutes**: es una variable numérica, que presenta una distribucion log-normal, con valores más acumulados cercano a 0. Posee una alta varianza en sus datos y no presenta valores nulos . En cuanto a los valores outliers, se puede apreciar que sí es una variable que los presenta debido a que tiene valores que están en la cola superior de los datos. En cuando a las correlaciones entre los datos, se puede observar que las que presentan mayor correlación no superan el valor de 0.2.

* **device_os**: esta variable tiene 5 valores posibles, other, linux, windows, macintosh y x11. Donde la calse mayoritaria, que es other presenta un 37% de los datos. Esta variable no tiene valores nulos y no presenta correlaciones entre otras variables.

* **keep_alive_session**: es una variable que presenta 2 valores posibles, 1 y 0, donde la clase mayoritaria es 1 que corresponde al 53% de los datos. Esta variable no presenta valores nulos y tampoco presenta correlaciones con otroas variables.

* **month**: es una variable que no presenta una distribución clara, es algo oscilante donde en el mes 3 y 6 hay picos. En cuanto a las correlaciones que presenta con las otras variables, es hasta el momento la que presenta más correlaciones entre las variables. Esta variable presenta correlaciones negativas con las variables: zip_count_4w, velocity_6h, velocity_24h y  velocity_4w, donde la última presenta un valor de -0.8 de la otra variable. Tampoco presenta nulos la variable.

* **name_email_similarity**: la distribución de la variable no tiene una distribución clara, tiene cierto caracter sinusoidal donde se dispara en el valor de 1, donde se acumula la moda de la variable. En cuanto a las correlaciones que tiene esa variable con otras variables, se puede observar que no presenta correlaciones muy altas con otras variables, de hecho la más alta no supera el módulo de 0.2. Esta variable no presenta valores nulos.

* **date_of_birth_distinct_emails_4w**: esta variable parece tener una distribución bimodal con los 2 picos presentes en 5 y en 12. Presenta mucha variacion entre sus datos y en cuanto a la correlacion con otras variables, la variable customer age presenta una correlacion igual a -0.5 y otras correlaciones menor alrededores de 0.2 en modulo. ESta variable no presenta valores nulos.

* **email_is_free**: esta variable presenta 2 valores, 0 y 1 y su distribuciones son parecidas, alrededor de 50% cada una. En cuanto a las correlaciones con otras variables, no superan el modulo de 0.2. Esta variable no presenta valores nulos.

* **device_distinct_emails_8w**: los posibles valores de esta variable son 0, 1 y 2, siendo la clase mayoritaria la clase 1, siendo el 97% de los casos. En cuanto a las correlaciones entre las otras variables, no presenta correlaciones altas. Tampoco presenta valores nulos la variable.

* **phone_home_valid**: los posibles valores de esta variable son 2, que son 1 y 0. ambos tienen alrededor de 50% de los datos. En cuanto a las correlaciones que pueden existir con las otras variables, no presenta altas correlaciones, siendo la mayoria bajo al modulo de 0.2.

* **phone_mobile_valid**: los posibles valores de esta variable son 1 y 0. donde la clase mayoritaria es 1 que está presente en el 89% de los datos. En cuando a las correlaciones entre las otras variables, no presenta correlaciones fuertes con las otras, de hecho presenta valores menores del modelo de 0.2.

* **segmentacion_etaria**: presenta 4 valores posibles donde la clase mayoritaria presenta el 71% de los datos. En cuanto a la correlacion entre otras variables, tiene sentido q esté altamente correlacionado con el customer_age ya que se construyó a partir de esa variable esa variable. No presenta variables nulas.








## 4. Preprocesamiento

### 4.1 Declarar `ColumnTransformer`

Dada la naturaleza de los datos, se procede a realizar un *standar scaler* para las variables numéricas y un *one hot encoder* para las categóricas. Esta decisión se toma en base a que luego debe reducirse la dimensionalidad y *ss* suele funcionar bien a la par de algoritmos como PCA. Por su parte utilizar *ohe* se utiliza para que las variables categóricas tengan sentido a la hora de realizar los reducción de dimensionalidad/modelos

In [48]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, MinMaxScaler, PowerTransformer

In [49]:
import sklearn

In [50]:
bbdd_final.columns

Index(['id', 'fraud_bool', 'income', 'current_address_months_count',
       'customer_age', 'days_since_request', 'payment_type', 'zip_count_4w',
       'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w',
       'employment_status', 'credit_risk_score', 'housing_status',
       'bank_months_count', 'has_other_cards', 'proposed_credit_limit',
       'foreign_request', 'source', 'session_length_in_minutes', 'device_os',
       'keep_alive_session', 'month', 'name_email_similarity',
       'date_of_birth_distinct_emails_4w', 'email_is_free',
       'device_distinct_emails_8w', 'phone_home_valid', 'phone_mobile_valid',
       'segmentacion_etaria'],
      dtype='object')

In [51]:
variables_numericas=[
    'income',
    'current_address_months_count',
    'customer_age',
    'days_since_request',
    'zip_count_4w',
    'velocity_6h',
    'velocity_24h',
    'velocity_4w',
    'bank_branch_count_8w',
    'credit_risk_score',
    'bank_months_count',
    'proposed_credit_limit',
    'session_length_in_minutes',
    'month',
    'name_email_similarity',
    'date_of_birth_distinct_emails_4w'
]
variables_ordinales=[
    'payment_type',
    'employment_status',
    'housing_status',
    'has_other_cards',
    'foreign_request',
    'source',
    'device_os',
    'keep_alive_session',
    'email_is_free',
    'device_distinct_emails_8w',
    'phone_home_valid',
    'phone_mobile_valid',
    'segmentacion_etaria'
] 

In [52]:
numerical_transformer = StandardScaler()
ordinal_transformer = OneHotEncoder(sparse_output=False)

In [53]:
#Realizamos One Hot Encoder a las variablees ordinales
column_transformer = ColumnTransformer(
    transformers=[ 
      ('ord_transform',ordinal_transformer, variables_ordinales),
      ('num_transform',numerical_transformer, variables_numericas),
      ('no_transform', 'passthrough', ['id', 'fraud_bool'])     
    ],
    remainder='passthrough'
)

### 4.2 Transformar datos

In [54]:
transformed_data = column_transformer.fit_transform(bbdd_final)

In [55]:
new_columns = column_transformer.named_transformers_['ord_transform'].get_feature_names_out(variables_ordinales)
new_columns_2 = column_transformer.named_transformers_['num_transform'].get_feature_names_out(variables_numericas)
all_columns = list(new_columns) + list(new_columns_2) + ['id', 'fraud_bool']

df_encoded = pd.DataFrame(transformed_data, columns=all_columns)

In [56]:
df_encoded.describe()

payment_type_AA  payment_type_AB  payment_type_AC  payment_type_AD  \
count    873770.000000     873770.00000    873770.000000    873770.000000   
mean          0.249429          0.39908         0.247095         0.104159   
std           0.432683          0.48971         0.431323         0.305467   
min           0.000000          0.00000         0.000000         0.000000   
25%           0.000000          0.00000         0.000000         0.000000   
50%           0.000000          0.00000         0.000000         0.000000   
75%           0.000000          1.00000         0.000000         0.000000   
max           1.000000          1.00000         1.000000         1.000000   

       payment_type_AE  employment_status_CA  employment_status_CB  \
count    873770.000000         873770.000000         873770.000000   
mean          0.000237              0.685136              0.135139   
std           0.015390              0.464462              0.341872   
min           0.000000              0.000000              0.000000   
25%           0.000000              0.000000              0.000000   
50%           0.000000              1.000000              0.000000   
75%           0.000000              1.000000              0.000000   
max           1.000000              1.000000              1.000000   

       employment_status_CC  employment_status_CD  employment_status_CE  \
count         873770.000000         873770.000000         873770.000000   
mean               0.088922              0.026679              0.018702   
std                0.284631              0.161143              0.135469   
min                0.000000              0.000000              0.000000   
25%                0.000000              0.000000              0.000000   
50%                0.000000              0.000000              0.000000   
75%                0.000000              0.000000              0.000000   
max                1.000000              1.000000              1.000000   

       employment_status_CF  employment_status_CG  housing_status_BA  \
count         873770.000000         873770.000000      873770.000000   
mean               0.044928              0.000496           0.215611   
std                0.207147              0.022256           0.411245   
min                0.000000              0.000000           0.000000   
25%                0.000000              0.000000           0.000000   
50%                0.000000              0.000000           0.000000   
75%                0.000000              0.000000           0.000000   
max                1.000000              1.000000           1.000000   

       housing_status_BB  housing_status_BC  housing_status_BD  \
count      873770.000000      873770.000000      873770.000000   
mean            0.301064           0.335327           0.026490   
std             0.458721           0.472105           0.160587   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           1.000000           0.000000   
max             1.000000           1.000000           1.000000   

       housing_status_BE  housing_status_BF  housing_status_BG  \
count      873770.000000      873770.000000      873770.000000   
mean            0.119589           0.001651           0.000268   
std             0.324480           0.040605           0.016363   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             0.000000           0.000000           0.000000   
max             1.000000           1.000000           1.000000   

       has_other_cards_0  has_other_cards_1  foreign_request_0  \
count      873770.000000      873770.000000      873770.000000   
mean            0.750619           

In [57]:
df_encoded.columns.values

array(['payment_type_AA', 'payment_type_AB', 'payment_type_AC',
       'payment_type_AD', 'payment_type_AE', 'employment_status_CA',
       'employment_status_CB', 'employment_status_CC',
       'employment_status_CD', 'employment_status_CE',
       'employment_status_CF', 'employment_status_CG',
       'housing_status_BA', 'housing_status_BB', 'housing_status_BC',
       'housing_status_BD', 'housing_status_BE', 'housing_status_BF',
       'housing_status_BG', 'has_other_cards_0', 'has_other_cards_1',
       'foreign_request_0', 'foreign_request_1', 'source_INTERNET',
       'source_TELEAPP', 'device_os_linux', 'device_os_macintosh',
       'device_os_other', 'device_os_windows', 'device_os_x11',
       'keep_alive_session_0', 'keep_alive_session_1',
       'email_is_free_0.0', 'email_is_free_1.0',
       'device_distinct_emails_8w_0.0', 'device_distinct_emails_8w_1.0',
       'device_distinct_emails_8w_2.0', 'phone_home_valid_0.0',
       'phone_home_valid_1.0', 'phone_mobile_valid_0

In [58]:
df_encoded['fraud_bool']= df_encoded['fraud_bool'].astype(int)

## 5. Visualización en Baja Dimensionalidad

### 5.1 Muestrear dataframe

Se procede a realizar una muestra del dataframe, en principio se pendó en que la muestra fuera aleatoria, pues esto conservaría las proporciones. Sin embargo, dado el desbalance de la clase de interés *fraud_bool*, se procede a realizar un muestreo a mano, para suplir en la medida de lo posible este desbalance, pero manteniendo la proporción por rango estario.

In [59]:
from sklearn.decomposition import PCA

In [60]:
menor_df = df_encoded[df_encoded['segmentacion_etaria_Menor a 18']>0.1]
joven_df = df_encoded[df_encoded['segmentacion_etaria_19 - 26 años']>0.1]
adulto_df = df_encoded[df_encoded['segmentacion_etaria_27 - 59 años']>0.1]
mayor_df = df_encoded[df_encoded['segmentacion_etaria_60 años o mas']>0.1]

In [61]:
prop_menor=np.round(menor_df.shape[0]/df_encoded.shape[0]*25000)
prop_joven=np.round(joven_df.shape[0]/df_encoded.shape[0]*25000)
prop_adulto=np.round(adulto_df.shape[0]/df_encoded.shape[0]*25000)
prop_mayor=np.round(mayor_df.shape[0]/df_encoded.shape[0]*25000)
print(prop_menor, prop_joven, prop_adulto, prop_mayor)

325.0 3673.0 17940.0 3062.0


Los valores presentes son las cantidades de datos que se extraerán de cada segmento etario, se seleccionarán hasta la mitad de datos con *fraud_bool* y el resto al azar entre los datos que no lo tienen. Luego se unen en una única tabla muestreo, para realizar la reducción de dimensionalidad

In [62]:
#fraudes
fraud_menor = menor_df[menor_df['fraud_bool']>0.1]
fraud_joven = joven_df[joven_df['fraud_bool']>0.1]
fraud_adulto =adulto_df[adulto_df['fraud_bool']>0.1]
fraud_mayor = mayor_df[mayor_df['fraud_bool']>0.1]

#no_fraudes
no_fraud_menor = menor_df[menor_df['fraud_bool']<=0.1]
no_fraud_joven = joven_df[joven_df['fraud_bool']<=0.1]
no_fraud_adulto =adulto_df[adulto_df['fraud_bool']<=0.1]
no_fraud_mayor = mayor_df[mayor_df['fraud_bool']<=0.1]

In [63]:
print(f'fraude menor: {fraud_menor.shape[0]} \nfraude joven: {fraud_joven.shape[0]} \nfraude adulto: {fraud_adulto.shape[0]} \nfraude adulto mayor: {fraud_mayor.shape[0]}')

fraude menor: 48 
fraude joven: 817 
fraude adulto: 7027 
fraude adulto mayor: 1730


Dados los volúmenes de fraudes, se proceden a dejar todos y se arma el muestreo aleatorio del resto de variables para alcanzar los valores esperaros.

In [64]:
muestreo_menor = no_fraud_menor.sample(n=int(prop_menor-fraud_menor.shape[0]), replace=True, random_state=190423)
muestreo_joven = no_fraud_joven.sample(n=int(prop_joven-fraud_joven.shape[0]), replace=True, random_state=190423)
muestreo_adulto = no_fraud_adulto.sample(n=int(prop_adulto-fraud_adulto.shape[0]), replace=True, random_state=190423)
muestreo_mayor = no_fraud_mayor.sample(n=int(prop_mayor-fraud_mayor.shape[0]), replace=True, random_state=190423)

In [65]:
muestreo = pd.concat([muestreo_menor, muestreo_joven, muestreo_mayor, muestreo_adulto, fraud_menor, fraud_joven, fraud_adulto, fraud_mayor], ignore_index=True)

In [66]:
muestreo.shape

(25000, 63)

Para realizar la reducción de dimensionalidad se opta por utilizar PCA

In [67]:
caract = df_encoded.columns.drop('fraud_bool', 'id')
pca = PCA(n_components=9)
transformed_data = pca.fit_transform(muestreo[caract])

In [68]:
transformed_data.shape

(25000, 9)

In [69]:
muestreo['comp1'] = transformed_data[:,0]
muestreo['comp2'] = transformed_data[:,1]
muestreo['comp3'] = transformed_data[:,2]
muestreo['comp4'] = transformed_data[:,3]
muestreo['comp5'] = transformed_data[:,4]
muestreo['comp6'] = transformed_data[:,5]
muestreo['comp7'] = transformed_data[:,6]
muestreo['comp8'] = transformed_data[:,7]
muestreo['comp9'] = transformed_data[:,8]

In [70]:
clases=['segmentacion_etaria_19 - 26 años',
       'segmentacion_etaria_27 - 59 años', 'segmentacion_etaria_60 años o mas',
       'segmentacion_etaria_Menor a 18']

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

In [71]:
menor = muestreo[muestreo['segmentacion_etaria_Menor a 18']>0][['comp1','comp2', 'fraud_bool']]
joven = muestreo[muestreo['segmentacion_etaria_19 - 26 años']>0][['comp1','comp2', 'fraud_bool']]
adulto = muestreo[muestreo['segmentacion_etaria_27 - 59 años']>0][['comp1','comp2', 'fraud_bool']]
mayor = muestreo[muestreo['segmentacion_etaria_60 años o mas']>0][['comp1','comp2', 'fraud_bool']]

### 5.3 Visualizar según rangos etarios

In [72]:
import plotly.express as px

fig = px.scatter(menor, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [73]:
fig = px.scatter(joven, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [74]:
fig = px.scatter(adulto, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [75]:
fig = px.scatter(mayor, x='comp1', y='comp2', color='fraud_bool')
fig.show()

Si bien no se pueden apreciar patrones a simple vista, si se pueden notar ciertas semejanzas o percepción de los datos.
* Para los menores, los fraudes parecen no tener un patrón seguro, además de tener un volumen menor, sin embargo si ponemos atención parece que una función sinusoidal funcionaría como aproximación a este.
* Para adultos y adultos mayores, se puede apreciar una leve translación en la segunda componente, es decir, los no-fraudes poseen un valor mayor en esta variable, en comparación a los fraudes.

## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

La forma en que se detectaran anomalías por rango etario serán usando modelos de clasificación. Se probarán base dummy, decision tree y knn. Los modelos que presenten un mejor recall y precisión serán los escogidos, dada la naturaleza del problema. Esto debido a que el recall es la proporción de la clase positiva que efectivamente fueron detectedas con éxito, debido al desbalance de clases existente, sería la mejor métrica a considerar.

In [76]:
from sklearn.dummy import DummyClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.svm import OneClassSVM
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

In [77]:
from sklearn.metrics import f1_score, recall_score, precision_score
from sklearn.model_selection import train_test_split
import numpy as np

def run_classifier(clf, X, y):
        metrics = {'f1-score': [], 'precision': [], 'recall': []}
    
        #dividiendo los datos de entrenamiento y validación
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30,  random_state=101)

        clf.fit(X_train,y_train) #fiteamos el modelo
        predictions = clf.predict(X_test) #Generamos las predicciones
       
        
        metrics['y_pred'] = predictions
        metrics['f1-score'].append(f1_score(y_test, predictions)) 
        metrics['recall'].append(recall_score(y_test, predictions))
        metrics['precision'].append(precision_score(y_test, predictions))
        matriz= confusion_matrix(y_test, predictions)


        return metrics, matriz

#### Menor

In [79]:
X=menor[['comp1','comp2']]
y=menor['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified', random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))

classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

----------------
Resultados para clasificador:  Base Dummy
[[74 11]
 [12  1]]
Precision promedio: 0.08333333333333333
Recall promedio: 0.07692307692307693
F1-score promedio: 0.08
----------------


----------------
Resultados para clasificador:  Decision Tree
[[81  4]
 [12  1]]
Precision promedio: 0.2
Recall promedio: 0.07692307692307693
F1-score promedio: 0.1111111111111111
----------------


----------------
Resultados para clasificador:  KNN
[[85  0]
 [13  0]]
Precision promedio: 0.0
Recall promedio: 0.0
F1-score promedio: 0.0
----------------




/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.



Acá se elige el modelo decision tree, para los menores.

In [80]:
X=menor[['comp1','comp2']]
y=menor['fraud_bool']

X_train_menor, X_test_menor, y_train_menor, y_test_menor = train_test_split(X, y, test_size=.30,  random_state=101)
clf_menor=DecisionTreeClassifier(max_depth=5)
clf_menor.fit(X_train_menor,y_train_menor) #fiteamos el modelo
predictions_menor = clf_menor.predict(X_test_menor) #Generamos las predicciones
predictions_menor_s=pd.DataFrame(predictions_menor).reset_index()
menor_2 =  pd.concat([X_test_menor.reset_index(), predictions_menor_s], axis=1)
confusion_matrix(y_test_menor, predictions_menor)

array([[82,  3],
       [12,  1]])

#### Joven

In [81]:
X=joven[['comp1','comp2']]
y=joven['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified', random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=10))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))

classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

----------------
Resultados para clasificador:  Base Dummy
[[653 187]
 [204  58]]
Precision promedio: 0.23673469387755103
Recall promedio: 0.22137404580152673
F1-score promedio: 0.2287968441814596
----------------


----------------
Resultados para clasificador:  Decision Tree
[[806  34]
 [254   8]]
Precision promedio: 0.19047619047619047
Recall promedio: 0.030534351145038167
F1-score promedio: 0.052631578947368425
----------------


----------------
Resultados para clasificador:  KNN
[[834   6]
 [261   1]]
Precision promedio: 0.14285714285714285
Recall promedio: 0.003816793893129771
F1-score promedio: 0.007434944237918214
----------------




Acá se elige el modelo de base dummy

In [82]:
X=joven[['comp1','comp2']]
y=joven['fraud_bool']

X_train_joven, X_test_joven, y_train_joven, y_test_joven = train_test_split(X, y, test_size=.30,  random_state=101)
clf_joven=DummyClassifier(strategy='stratified', random_state=10)
clf_joven.fit(X_train_joven,y_train_joven) #fiteamos el modelo
predictions_joven = clf_joven.predict(X_test_joven) #Generamos las predicciones
predictions_joven_s=pd.DataFrame(predictions_joven).reset_index()
joven_2 =  pd.concat([X_test_joven.reset_index(), predictions_joven_s], axis=1)
confusion_matrix(y_test_joven, predictions_joven)

array([[653, 187],
       [204,  58]])

#### Adulto

In [83]:
# Ejecutar este código

X=adulto[['comp1','comp2']]
y=adulto['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified',  random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c3 = ("KNN", KNeighborsClassifier(n_neighbors=10))


classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

----------------
Resultados para clasificador:  Base Dummy
[[2040 1314]
 [1251  777]]
Precision promedio: 0.3715925394548063
Recall promedio: 0.3831360946745562
F1-score promedio: 0.37727603787327024
----------------


----------------
Resultados para clasificador:  Decision Tree
[[2997  357]
 [1584  444]]
Precision promedio: 0.5543071161048689
Recall promedio: 0.21893491124260356
F1-score promedio: 0.3138918345705196
----------------


----------------
Resultados para clasificador:  KNN
[[2813  541]
 [1701  327]]
Precision promedio: 0.37672811059907835
Recall promedio: 0.16124260355029585
F1-score promedio: 0.22582872928176795
----------------




Acá se elige el modelo de base dummy

In [84]:
X=adulto[['comp1','comp2']]
y=adulto['fraud_bool']

X_train_adulto, X_test_adulto, y_train_adulto, y_test_adulto = train_test_split(X, y, test_size=.30,  random_state=101)
clf_adulto=DummyClassifier(strategy='stratified', random_state=10)
clf_adulto.fit(X_train_adulto,y_train_adulto) #fiteamos el modelo
predictions_adulto = clf_adulto.predict(X_test_adulto) #Generamos las predicciones
predictions_adulto_s=pd.DataFrame(predictions_adulto).reset_index()
adulto_2 =  pd.concat([X_test_adulto.reset_index(), predictions_adulto_s], axis=1)
confusion_matrix(y_test_adulto, predictions_adulto)

array([[2040, 1314],
       [1251,  777]])

#### Mayor

In [85]:
# Ejecutar este código

X=mayor[['comp1','comp2']]
y=mayor['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified',  random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))


classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

----------------
Resultados para clasificador:  Base Dummy
[[190 223]
 [207 299]]
Precision promedio: 0.5727969348659003
Recall promedio: 0.5909090909090909
F1-score promedio: 0.5817120622568093
----------------


----------------
Resultados para clasificador:  Decision Tree
[[162 251]
 [106 400]]
Precision promedio: 0.6144393241167435
Recall promedio: 0.7905138339920948
F1-score promedio: 0.6914433880726016
----------------


----------------
Resultados para clasificador:  KNN
[[187 226]
 [221 285]]
Precision promedio: 0.5577299412915852
Recall promedio: 0.5632411067193676
F1-score promedio: 0.56047197640118
----------------




Acá se elige el modelo decision tree, pues para adultos mayores parece tener un buen recall y precisión.

In [86]:
X=mayor[['comp1','comp2']]
y=mayor['fraud_bool']

X_train_mayor, X_test_mayor, y_train_mayor, y_test_mayor = train_test_split(X, y, test_size=.30,  random_state=101)
clf_mayor=DecisionTreeClassifier(max_depth=5)
clf_mayor.fit(X_train_mayor,y_train_mayor) #fiteamos el modelo
predictions_mayor = clf_mayor.predict(X_test_mayor) #Generamos las predicciones
predictions_mayor_s=pd.DataFrame(predictions_mayor).reset_index()
mayor_2 =  pd.concat([X_test_mayor.reset_index(), predictions_mayor_s], axis=1)
confusion_matrix(y_test_mayor, predictions_mayor)

array([[162, 251],
       [106, 400]])

### 6.2 Agregar resultados a dataframe de muestreo




En la base de datos que se construirá, se dejarán los valores de la matriz de confusión de cada uno de los segmentos con el modelo dummy.

In [87]:
# initialize list of lists
data = [['menor',82, 3, 12, 1], ['joven', 653, 187, 204, 58], ['adulto', 2040,1314,1251,777],['mayor',160,253,105,401]]
  
# Create the pandas DataFrame
df = pd.DataFrame(data, columns=['segmento', 'TP', 'FP','FN','TN'])
  
# print dataframe.
df

segmento    TP    FP    FN   TN
0    menor    82     3    12    1
1    joven   653   187   204   58
2   adulto  2040  1314  1251  777
3    mayor   160   253   105  401

### 6.3 Visualizar según rangos etarios

#### Menor

In [88]:
fig = px.scatter(menor_2, x='comp1', y='comp2', color=0)
fig.show()

####Joven

In [89]:
fig = px.scatter(joven_2, x='comp1', y='comp2', color=0)
fig.show()

####Adulto

In [90]:
fig = px.scatter(adulto_2, x='comp1', y='comp2', color=0)
fig.show()

####Mayor

In [91]:
fig = px.scatter(mayor_2, x='comp1', y='comp2', color=0)
fig.show()

### 6.4 Calcular ratios y responder

Finalmente calcularemos un par de ratios para concluir con la evaluación de los modelos.
* P: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude.
* R: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude.

In [92]:
pred_fraude = (df['TP'] + df['FP']).sum()
real_fraude = (df['TP'] + df['FN']).sum()

In [93]:
P = df['TP'].sum() / pred_fraude
P

0.625532821824382

In [94]:
R = df['TP'].sum() / real_fraude
R

0.6512092300865321

# Conclusiones

Dados los ratios, se pueden apreciar que la detección tuvo un éxtito cercano al 65% para todo el dataset. Si nos vamos de manera particular, resultó mucho más fácil predecir cuando existían muchos datos, como por ejemplo para adultos o adultos mayores, que para jóvenes y menores de edad.

Los modelos implementados no poseen un buen desempreño del todo, quizás se produjeron problemas al momento de realizar la reducción de dimensionalidad, o también puede ser que el haber aplicado el *standar scaller* a todos los datos por igual no haya sido el mejor acercamiento, pues había una o dos columnas con muchos outlayers que se puedieron haber pasado por un *Power Transformer* para que fueran más manejables a la hora de realizar el PCA. 

A pesar de lo anterior, obtener un 65% de desempeño se considera correcto antes la versión *naive* de modelado, pues no se probaron distintos hiperparámetros para lograr un mejor desempeño. Actividad que sería altamente recomendada en caso de querer seguir con el proyecto y detectar automáticamente fraudes en un futuro de manera más concisa, a pesar de esto, puede ser útil para reducir considerablemente la cantidad de casos que se tienen en la mira por posible fraude.

Finalmente, se recalca el hecho de que es necesario entender la data para que los modelos tengas un funcionamiento correcto.